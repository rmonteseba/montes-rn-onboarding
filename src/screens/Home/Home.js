import React, { useEffect, useState } from 'react';
import { Alert, View, StyleSheet } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import { getPopularMovies, getTopRatedMovies } from '@/actions/MovieActions';
import { getPopularMoviesState, getTopRatedMoviesState } from '@/selectors/MovieSelectors';
import MovieCarousel from '@/components/MovieCarousel';
import { strings } from '@/localization';

const baseStyles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  carousel: {
    base: {
      marginHorizontal: 20,
    },
  },
});

const Home = () => {
  const dispatch = useDispatch();
  const { fulfilled: topRatedMoviesFulfilled, error: topRatedMoviesError } =
    useSelector(getTopRatedMoviesState);

  const { fulfilled: popularMoviesFulfilled, error: popularMoviesError } =
    useSelector(getPopularMoviesState);

  const [page, setPage] = useState(1);

  useEffect(() => {
    dispatch(getTopRatedMovies({ page }));
    dispatch(getPopularMovies({ page }));
  }, [page, dispatch]);

  useEffect(() => {
    if (topRatedMoviesError) {
      Alert.alert(topRatedMoviesError); // TODO improve error message :)
    }
  }, [topRatedMoviesError]);

  useEffect(() => {
    if (popularMoviesError) {
      Alert.alert(popularMoviesError); // TODO improve error message :)
    }
  }, [popularMoviesError]);

  return (
    <View style={baseStyles.container}>
      <MovieCarousel
        styles={baseStyles.carousel.base}
        titleStyles={baseStyles.carousel.title}
        title={strings.home.popularMovies}
        movies={popularMoviesFulfilled || []}
      />
      <MovieCarousel
        styles={baseStyles.carousel.base}
        titleStyles={baseStyles.carousel.title}
        title={strings.home.topFeaturedMovies}
        movies={topRatedMoviesFulfilled || []}
      />
    </View>
  );
};

export default Home;
