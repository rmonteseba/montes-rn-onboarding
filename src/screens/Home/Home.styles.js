import { StyleSheet } from 'react-native';

export const baseStyles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  carousel: {
    base: {
      marginHorizontal: 20,
    },
  },
});
