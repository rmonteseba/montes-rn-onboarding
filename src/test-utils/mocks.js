export const movie = {
  adult: false,
  backdrop_path: '/fBshLiEJcjdfrU3qQBIINcePSsm.jpg',
  belongs_to_collection: null,
  budget: 0,
  genres: [
    { id: 16, name: 'Animation' },
    { id: 878, name: 'Science Fiction' },
    { id: 12, name: 'Adventure' },
    { id: 10751, name: 'Family' },
  ],
  homepage: '',
  id: 877269,
  imdb_id: 'tt10298840',
  original_language: 'en',
  original_title: 'Strange World',
  overview:
    'A journey deep into an uncharted and treacherous land, where fantastical creatures await the legendary Clades—a family of explorers whose differences threaten to topple their latest, and by far most crucial, mission.',
  popularity: 63.426,
  poster_path: '/a9x70QUXs6uEwtin1SGNKoFXoia.jpg',
  production_companies: [
    {
      id: 2,
      logo_path: '/wdrCwmRnLFJhEoH8GSfymY85KHT.png',
      name: 'Walt Disney Pictures',
      origin_country: 'US',
    },
    {
      id: 6125,
      logo_path: '/tzsMJBJZINu7GHzrpYzpReWhh66.png',
      name: 'Walt Disney Animation Studios',
      origin_country: 'US',
    },
  ],
  production_countries: [{ iso_3166_1: 'US', name: 'United States of America' }],
  release_date: '2022-11-23',
  revenue: 0,
  runtime: 102,
  spoken_languages: [
    { english_name: 'English', iso_639_1: 'en', name: 'English' },
    { english_name: 'Slovak', iso_639_1: 'sk', name: 'Slovenčina' },
  ],
  status: 'In Production',
  tagline: 'Journey to a place where nothing is as it appears!',
  title: 'Strange World',
  video: false,
  vote_average: 0.0,
  vote_count: 0,
};

export const movieResponse = movie;

export const movies = [
  {
    adult: false,
    backdrop_path: '/rSPw7tgCH9c6NqICZef4kZjFOQ5.jpg',
    genre_ids: [18, 80],
    id: 238,
    original_language: 'en',
    original_title: 'The Godfather',
    overview:
      'Spanning the years 1945 to 1955, a chronicle of the fictional Italian-American Corleone crime family. When organized crime family patriarch, Vito Corleone barely survives an attempt on his life, his youngest son, Michael steps in to take care of the would-be killers, launching a campaign of bloody revenge.',
    popularity: 105.925,
    poster_path: '/3bhkrj58Vtu7enYsRolD1fZdja1.jpg',
    release_date: '1972-03-14',
    title: 'The Godfather',
    video: false,
    vote_average: 8.7,
    vote_count: 16846,
  },
  {
    adult: false,
    backdrop_path: '/kXfqcdQKsToO0OUXHcrrNCHDBzO.jpg',
    genre_ids: [18, 80],
    id: 278,
    original_language: 'en',
    original_title: 'The Shawshank Redemption',
    overview:
      'Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.',
    popularity: 88.027,
    poster_path: '/q6y0Go1tsGEsmtFryDOJo3dEmqu.jpg',
    release_date: '1994-09-23',
    title: 'The Shawshank Redemption',
    video: false,
    vote_average: 8.7,
    vote_count: 22585,
  },
  {
    adult: false,
    backdrop_path: '/aVFx1VtlOxR3v0ADEatalXOvwbu.jpg',
    genre_ids: [16, 14, 28],
    id: 620249,
    original_language: 'zh',
    original_title: '罗小黑战记',
    overview:
      'In the bustling human world, spirits live peacefully with mankind. Luo Xiaohei, a young cat spirit, becomes a nomad after his forest home is destroyed. His natural talent soon attracts spirits and humans alike, but all of them have their own motivations.',
    popularity: 19.13,
    poster_path: '/aLv87NgRJUPkQ6sVLP72IisDdt4.jpg',
    release_date: '2019-08-27',
    title: 'The Legend of Hei',
    video: false,
    vote_average: 8.7,
    vote_count: 200,
  },
];

export const moviesResponse = {
  page: 1,
  results: movies,
  total_pages: 524,
  total_results: 10469,
};
