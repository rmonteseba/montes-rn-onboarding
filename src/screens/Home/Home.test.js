import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import Home from '@/screens/Home/Home';
import { withProviders } from '@/test-utils';
import { moviesResponse } from '@/test-utils/mocks';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

const fakeStore = {
  error: {},
  status: {},
  user: {
    username: 'johndoe',
  },
};

describe('Home', () => {
  let axiosMock;

  beforeEach(() => {
    axiosMock = new MockAdapter(axios);
    axiosMock.onGet(`/movie/upcoming`).reply(200, moviesResponse);
    axiosMock.onGet(`/movie/top_rated`).reply(200, moviesResponse);
    axiosMock.onGet(`/movie/popular`).reply(200, moviesResponse);
  });

  afterEach(() => {
    axiosMock.restore();
  });

  it('should show the movie list of the main carousel', () => {
    const { getAllByTestId, toJSON } = render(withProviders(<Home />, { initialState: fakeStore }));
    // expect(getAllByTestId('movie-card').length).toBe(moviesResponse.results.length); TODO solve items rendering on carousel
    expect(toJSON()).toMatchSnapshot();
  });

  it('should browse to the detail page', () => {
    const { getByTestId } = render(withProviders(<Home />, { initialState: fakeStore }));
    const movieDetailButton = getByTestId('movie-detail-btn');

    fireEvent.press(movieDetailButton);
    expect(mockedNavigate).toHaveBeenCalled();
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(withProviders(<Home />, { initialState: fakeStore }));
    expect(toJSON()).toMatchSnapshot();
  });
});
