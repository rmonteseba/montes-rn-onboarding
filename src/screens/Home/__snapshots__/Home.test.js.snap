// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match the snapshot 1`] = `
<RCTScrollView
  collapsable={false}
  handlerTag={15}
  handlerType="NativeViewGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  waitFor={
    Array [
      Object {
        "current": null,
      },
    ]
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View>
        <View>
          <View
            collapsable={false}
            handlerTag={16}
            handlerType="PanGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "height": 450,
                  "width": 750,
                },
                Array [
                  Object {
                    "height": 450,
                    "width": 750,
                  },
                  undefined,
                ],
                Object {
                  "flexDirection": "row",
                },
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "bottom": -45,
              "display": "flex",
              "flexDirection": "row",
              "position": "absolute",
            }
          }
        >
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={17}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            touchSoundDisabled={false}
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "padding": 50,
                  }
                }
              >
                <SvgMock
                  height={40}
                  width={40}
                />
                <Text
                  style={
                    Object {
                      "paddingVertical": 16,
                    }
                  }
                >
                  My List
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={18}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            touchSoundDisabled={false}
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "padding": 50,
                  }
                }
              >
                <SvgMock
                  height={40}
                  width={40}
                />
                <Text
                  style={
                    Object {
                      "paddingVertical": 16,
                    }
                  }
                >
                  Play
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={19}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="movie-detail-btn"
            touchSoundDisabled={false}
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "padding": 50,
                  }
                }
              >
                <SvgMock
                  height={40}
                  width={40}
                />
                <Text
                  style={
                    Object {
                      "paddingVertical": 16,
                    }
                  }
                >
                  Info
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 25,
                "fontWeight": "bold",
                "marginLeft": 40,
                "marginVertical": 15,
              },
              undefined,
            ]
          }
        >
          Popular Movies
        </Text>
        <View
          collapsable={false}
          handlerTag={20}
          handlerType="PanGestureHandler"
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "height": 180,
                "width": 188,
              },
              Array [
                Object {
                  "height": 180,
                  "width": 750,
                },
                Object {
                  "marginHorizontal": 20,
                },
              ],
              Object {
                "flexDirection": "row",
              },
            ]
          }
        />
      </View>
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 25,
                "fontWeight": "bold",
                "marginLeft": 40,
                "marginVertical": 15,
              },
              undefined,
            ]
          }
        >
          Top Featured
        </Text>
        <View
          collapsable={false}
          handlerTag={21}
          handlerType="PanGestureHandler"
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "height": 180,
                "width": 188,
              },
              Array [
                Object {
                  "height": 180,
                  "width": 750,
                },
                Object {
                  "marginHorizontal": 20,
                },
              ],
              Object {
                "flexDirection": "row",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Home should show the movie list of the main carousel 1`] = `
<RCTScrollView
  collapsable={false}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  waitFor={
    Array [
      Object {
        "current": null,
      },
    ]
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View>
        <View>
          <View
            collapsable={false}
            handlerTag={2}
            handlerType="PanGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "height": 450,
                  "width": 750,
                },
                Array [
                  Object {
                    "height": 450,
                    "width": 750,
                  },
                  undefined,
                ],
                Object {
                  "flexDirection": "row",
                },
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "bottom": -45,
              "display": "flex",
              "flexDirection": "row",
              "position": "absolute",
            }
          }
        >
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={3}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            touchSoundDisabled={false}
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "padding": 50,
                  }
                }
              >
                <SvgMock
                  height={40}
                  width={40}
                />
                <Text
                  style={
                    Object {
                      "paddingVertical": 16,
                    }
                  }
                >
                  My List
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={4}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            touchSoundDisabled={false}
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "padding": 50,
                  }
                }
              >
                <SvgMock
                  height={40}
                  width={40}
                />
                <Text
                  style={
                    Object {
                      "paddingVertical": 16,
                    }
                  }
                >
                  Play
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={5}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="movie-detail-btn"
            touchSoundDisabled={false}
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "padding": 50,
                  }
                }
              >
                <SvgMock
                  height={40}
                  width={40}
                />
                <Text
                  style={
                    Object {
                      "paddingVertical": 16,
                    }
                  }
                >
                  Info
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 25,
                "fontWeight": "bold",
                "marginLeft": 40,
                "marginVertical": 15,
              },
              undefined,
            ]
          }
        >
          Popular Movies
        </Text>
        <View
          collapsable={false}
          handlerTag={6}
          handlerType="PanGestureHandler"
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "height": 180,
                "width": 188,
              },
              Array [
                Object {
                  "height": 180,
                  "width": 750,
                },
                Object {
                  "marginHorizontal": 20,
                },
              ],
              Object {
                "flexDirection": "row",
              },
            ]
          }
        />
      </View>
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 25,
                "fontWeight": "bold",
                "marginLeft": 40,
                "marginVertical": 15,
              },
              undefined,
            ]
          }
        >
          Top Featured
        </Text>
        <View
          collapsable={false}
          handlerTag={7}
          handlerType="PanGestureHandler"
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "height": 180,
                "width": 188,
              },
              Array [
                Object {
                  "height": 180,
                  "width": 750,
                },
                Object {
                  "marginHorizontal": 20,
                },
              ],
              Object {
                "flexDirection": "row",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
